data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/S3Service.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.sts.StsClient;
+
+public class S3Service {
+    private static final Logger LOG = LoggerFactory.getLogger(S3Service.class);
+
+    private S3SourceConfig s3SourceConfig;
+    private S3Client s3Client;
+
+    public S3Service(final S3SourceConfig s3SourceConfig) {
+        this.s3SourceConfig = s3SourceConfig;
+        this.s3Client = createS3Client(StsClient.create());
+    }
+
+    S3ObjectReference addS3Object(final S3ObjectReference s3ObjectReference) {
+        // TODO: should return message id and receipt handle if successfully converted to event
+        return null;
```

Comment:
I have a few concerns over this method and subsequent workflow. I have some ideas for improvements but I want to understand a few things first.

What are we adding an S3Object to?  I find this method name confusing and was really confused by the usage in the `run()` method.

Why are the parameters and return object the same object (S3ObjectReference)? Are we extending S3ObjectReference and expecting this class to mutate the variable with a messageId and Receipt? (I want to see if we can avoid mutation and nullable variables)

How is the receipt used by the calling function? Do we need it?

Reply:
> Why are the parameters and return object the same object (S3ObjectReference)? Are we extending S3ObjectReference and expecting this class to mutate the variable with a messageId and Receipt? (I want to see if we can avoid mutation and nullable variables)

@asifsmohammed and I discussed a couple approaches here:
1. We can make this templated and add something like `T getNotificationMetadata()` which in this case returns SQS metadata. It could be extended for SNS metadata if that were ever added.
2. We could extend this class with a sub-class.

In either case, the SQS metadata would include the `ReceiptHandle` and `MessageId`.

I'm fine with either approach. But, the inheritance approach would have the benefit of completely abstracting this metadata from the S3 code.

Reply:
My initial thought was to add `ReceiptHandle` and `MessageId` to `S3ObjectReference` which will be used to get S3 Object.
If S3 object is successfully processed we return the same `S3ObjectReference` to delete message from queue. 

I'll create a subclass which contains `ReceiptHandle` and `MessageId`.